### ISSUE TITLE
ability to execute shell command & capture stdout/err while process is running

### ISSUE THREAD

**Issue by despiegk** [vlang/v#6706]
We love the potential of vlang and are looking for some help around how to execute cmd's with access to sdout/err streams.

I didn't find a way how to execute a command and real-time process the stdout/error while the program is running.
Would like to use vlang to automate deployments & buildscripts but we need to be able to log whats going on while it gets executed.

In crystal lang we could use it as follows: see https://crystal-lang.org/api/0.34.0/Process/Redirect.html

Any idea, or is it already possible and we just didn't find it.

any help appreciated

regards

Kristof

**Comment by medvednikov** [comment]
Yes, this is something I also needed several times, and it's one of the missing crucial API features.

I'll work on the architecture and actual implementation this week, most likely it'll be similar to how Go does it.

**Comment by medvednikov** [comment]
I've implemented a basic proof of concept solution in 10 minutes:

```
import os

fn main() {
	mut cmd := os.Command{
		path:'cat'
		redirect_stdout: true
	}
	cmd.start()
	for !cmd.eof {
		line := cmd.read_line()
		println('line="$line"')
	}
	cmd.close()
}
```

The API will definitely improve :)

**Comment by despiegk** [comment]
thank you, amazing.
Would. it be easy to execute this in 1 thread and consume output in another thread?

**Comment by medvednikov** [comment]
Not at the moment, but @spytheman did some great work on this back in May. It's available in his fork: 

https://github.com/spytheman/v/commit/8e591b7f50bbc20b1a1b5a00998efe549327e0ea

This will be merged into the main tree soon.

**Comment by Dialga** [comment]
@spytheman's excellent work has been merged into mainline in PR #6786 and the POC was in commit 1b1d17cfb54b13702fbf5e6480d195d786354511. Check the docs https://modules.vlang.io/os.html#Process and https://modules.vlang.io/os.html#Command on how to use.
